include/onnxruntime/core/platform/threadpool.h

```diff
@@ -1,95 +1,267 @@
+  // REQUIRES: num_threads > 0
+  ThreadPool(Env* env, const ThreadOptions& thread_options, const NAME_CHAR_TYPE* name, int num_threads,
+             Eigen::Allocator* allocator);
+  // Constructs a pool that wraps around the thread::ThreadPoolInterface
+  // instance provided by the caller. Caller retains ownership of
+  // `user_threadpool` and must ensure its lifetime is longer than the
+  // ThreadPool instance.
+  ThreadPool(Eigen::ThreadPoolInterface* user_threadpool, Eigen::Allocator* allocator);
+
+  // Waits until all scheduled work has finished and then destroy the
+  // set of threads.
+  ~ThreadPool();
+
+  // Schedules fn() for execution in the pool of threads.
   void Schedule(std::function<void()> fn);
 
-  /*
-  Schedule work in the interval [0, total).
-  */
-  void ParallelFor(int32_t total, std::function<void(int32_t)> fn);
+  void SetStealPartitions(const std::vector<std::pair<unsigned, unsigned>>& partitions);
+
+  void ScheduleWithHint(std::function<void()> fn, int start, int limit);
+
+  // Returns the number of shards used by ParallelForFixedBlockSizeScheduling
+  // with these parameters.
+  int NumShardsUsedByFixedBlockSizeScheduling(int64_t total, int64_t block_size);
+
+  // Returns the number of threads spawned by calling TransformRangeConcurrently
+  // with these parameters.
+  // Deprecated. Use NumShardsUsedByFixedBlockSizeScheduling.
+  int NumShardsUsedByTransformRangeConcurrently(int64_t block_size, int64_t total);
+
+  // ParallelFor shards the "total" units of work assuming each unit of work
+  // having roughly "cost_per_unit" cost, in cycles. Each unit of work is
+  // indexed 0, 1, ..., total - 1. Each shard contains 1 or more units of work
+  // and the total cost of each shard is roughly the same.
+  //
+  // "cost_per_unit" is an estimate of the number of CPU cycles (or nanoseconds
+  // if not CPU-bound) to complete a unit of work. Overestimating creates too
+  // many shards and CPU time will be dominated by per-shard overhead, such as
+  // Context creation. Underestimating may not fully make use of the specified
+  // parallelism, and may also cause inefficiencies due to load balancing
+  // issues and stragglers.
+  void ParallelFor(std::ptrdiff_t total, double cost_per_unit,
+                   const std::function<void(std::ptrdiff_t first, std::ptrdiff_t)>& fn);
+  void ParallelFor(std::ptrdiff_t total, const TensorOpCost& cost_per_unit,
+                   const std::function<void(std::ptrdiff_t first, std::ptrdiff_t)>& fn);
+  // Similar to ParallelFor above, but takes the specified scheduling strategy
+  // into account.
+  void ParallelFor(int64_t total, const SchedulingParams& scheduling_params,
+                   const std::function<void(ptrdiff_t, ptrdiff_t)>& fn);
+
+  // Same as ParallelFor with Fixed Block Size scheduling strategy.
+  // Deprecated. Prefer ParallelFor with a SchedulingStrategy argument.
+  void TransformRangeConcurrently(int64_t block_size, int64_t total,
```

Comment:
>TransformRangeConcurrently [](start = 7, length = 26)

I don't see it is being available now. Should the existing code be refactored instead of introducing this and deprecating right away?

Reply:
Let me search. If it is not used, I'll delete it.